security/pkg/nodeagent/caclient/client.go

```diff
@@ -15,46 +15,87 @@
 package caclient
 
 import (
-	"crypto/x509"
-	"errors"
+	"encoding/base64"
 	"fmt"
 	"strings"
+	"time"
 
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials"
+	"k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
 
 	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/k8s/controller"
 	caClientInterface "istio.io/istio/security/pkg/nodeagent/caclient/interface"
+	citadel "istio.io/istio/security/pkg/nodeagent/caclient/providers/citadel"
 	gca "istio.io/istio/security/pkg/nodeagent/caclient/providers/google"
 )
 
-const googleCA = "GoogleCA"
+const (
+	googleCAName = "GoogleCA"
+	citadelName  = "Citadel"
+
+	retryInterval = time.Second * 2
+	maxRetries    = 100
+)
+
+type configMap interface {
+	GetCATLSRootCert() (string, error)
+}
 
 // NewCAClient create an CA client.
 func NewCAClient(endpoint, CAProviderName string, tlsFlag bool) (caClientInterface.Client, error) {
-	var opts grpc.DialOption
-	if tlsFlag {
-		pool, err := x509.SystemCertPool()
+	switch CAProviderName {
+	case googleCAName:
+		return gca.NewGoogleCAClient(endpoint, tlsFlag)
+	case citadelName:
+		cs, err := createClientSet()
 		if err != nil {
-			log.Errorf("could not get SystemCertPool: %v", err)
-			return nil, errors.New("could not get SystemCertPool")
+			return nil, err
 		}
-		creds := credentials.NewClientTLSFromCert(pool, "")
-		opts = grpc.WithTransportCredentials(creds)
-	} else {
-		opts = grpc.WithInsecure()
+		controller := controller.NewConfigMapController("", cs.CoreV1())
```

Comment:
why put empty namespace here ?

Reply:
This means the namespace the node agent is running in, right?